### PR TITLE
Move validation from widgets to the form fields

### LINKED ISSUE #441: Problem to validate short business numbers
Hi,

I just tried django-phonenumber-field in my project and it seems so good.
but I have problem to enter short numbers specially 4 or 5 digit numbers that used to use for business call centers.

Is there any settings that let me validate such numbers or it need a hack?

### LINKED ISSUE #597: Use Django's "normal" validation, not hardcoded in __init__()
This is a very cool project - but I need to "validate" my prhone numbers and clean them before the PhoneNumerField gets hand on it. It may have to do something with #441 and could be solved together with that issue too IMHO.

Use case: I have an input form that asks for a phone number, and it is likely one of two countries (DE,AT). People tend to fill in , numbers in a local format, and I want to call `clean_phone_number()` before the PhoneNumberField calls it's validators, and try to convert it into an international format.

ATM this is not possible, as the validation is done more or less hard coded. I can't override the form's  `clean_phone_number()` or  `clean()` methods, as PhoneNumberField does it's own validation stuff already before, and throws an error before my cleaning even starts.

Could you rearrange the code so that PhoneNumberField uses the default Django validation methods in the right order?
See https://docs.djangoproject.com/en/5.0/ref/forms/validation/

The validation should not be done in the `__init__()` method, but in the field's  `validate()`. This breaks the possibility of "normal" Django usage of this else very cool module.

I think this would close #441 too as we then could use custom validators.

### PR DESCRIPTION
**Pull request by francoisfreitag**
Move validation from widgets to the form fields

Previously, the widgets handled part of the validation. That behavior prevents overriding validation in form fields, as widgets were casting the value into a `PhoneNumber` object, validating it in the process.

Following the `MultiValueField` implementation from Django (and `MultiWidget`), the widget now handles the presentation logic, but makes no attempt at validation.

The new `SplitPhoneNumberField` handles the logic of validating the region choice and the number, and the `PhoneNumberPrefixWidget` simply dispatches the region and number data to the appropriate widget.

In order to retain backward compatibility, now that the `validate_international_phonenumber` actually comes into play, it’s error code has been changed to `invalid`, so that the custom error message for invalid shows.

Migration guide
===============

`validate_international_phonenumber`
------------------------------------

Review uses of the `invalid_phone_number` error code. Given that the validator usually did not come into play, you shouldn’t find many uses.

`PhoneNumberField` with `RegionalPhoneNumberWidget`
---------------------------------------------------

Make sure custom validation occurs in the Django `Form` `clean_FIELD()` (or `clean()`), and no changes should be noticeable.

`PhoneNumberField` with `PhoneNumberPrefixWidget`
-------------------------------------------------

Use the `SplitPhoneNumberField` instead. Error messages will change slightly and should be more precise (whether the region is not part of the choices, or the number cannot be interpreted in the selected region).

For more examples, take a look at `tests.test_formfields.SplitPhoneNumberFieldTest`. In particular `test_custom_attrs` and `test_custom_choices`, to see how they were migrated to the new field.

Finally, added a `validate_phonenumber` validator that allows short numbers.

Closes #441 
Closes #597

### CODE REVIEW

**Comment by francoisfreitag** [comment]
@stefanfoulis Would you like to review this PR?
If so, it can easily wait a few weeks. Otherwise, I would like to proceed with this change in an 8.0.0 release.

**Comment by amateja** [comment]
Hi @stefanfoulis,

I'm honored but recently I've been a bit stretched. I should find some time next week.

File: phonenumber_field/formfields.py [hunk 1]
```diff
@@ -1,19 +1,33 @@
...
+    import babel
+except ModuleNotFoundError:
+    babel = None  # type: ignore
+
+# ISO 3166-1 alpha-2 to national prefix
+REGION_CODE_TO_COUNTRY_CODE = {
+    region_code: country_code
+    for country_code, region_codes in COUNTRY_CODE_TO_REGION_CODE.items()
+    for region_code in region_codes
+}
```
[the remark this answers is not in the document]
Comment (stefanfoulis) [reply phonenumber_field/formfields.py:25]: `phonenumbers.COUNTRY_CODES_FOR_NON_GEO_REGIONS` seems to be exactly that list. Perhaps we should include those variants in our `REGION_CODE_TO_COUNTRY_CODE` as well :thinking: . But might be tricky to get that to correctly work with the widget and stuff.
```
In [13]: phonenumbers.COUNTRY_CODES_FOR_NON_GEO_REGIONS
Out[13]: {800, 808, 870, 878, 881, 882, 883, 888, 979}
```

[the remark this answers is not in the document]
Comment (stefanfoulis) [reply phonenumber_field/formfields.py:25]: Anyway... I'm not sure how the "old" implementation dealt with this. So if this is not a regression I don't think it is necessarily to fix in this PR.

**Review by stefanfoulis — APPROVED** [review]
:rocket:

File: phonenumber_field/formfields.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply phonenumber_field/formfields.py:25]: Thanks, I never noticed this entry, and that’s indeed a bug.
The entry shows up as “world” in the prefix `<select>`, with only `+979` (as you noticed). I opened https://github.com/stefanfoulis/django-phonenumber-field/issues/605 to deal with it. I’ll wait until this PR lands before touching that issue though.

**Pull request merged**